ctapipe_io_lst/calibration.py

```diff
@@ -0,0 +1,777 @@
+            )
+
+    def calibrate(self, event: ArrayEventContainer):
+        for tel_id in event.r0.tel:
+            r1 = event.r1.tel[tel_id]
+            waveform = r1.waveform
+
+            # do gain selection before converting to pe
+            # like eventbuilder will do
+            if self.select_gain:
+                selected_gain_channel = self.gain_selector(waveform)
+            else:
+                selected_gain_channel = None
+
+            # apply monitoring data corrections,
+            # subtract pedestal and convert to pe
+            if self.mon_data is not None:
+                calibration = self.mon_data.tel[tel_id].calibration
+                waveform -= calibration.pedestal_per_sample[:, :, np.newaxis]
+                waveform *= calibration.dc_to_pe[:, :, np.newaxis]
+
+            waveform = waveform.astype(np.float32)
+            n_gains, n_pixels, n_samples = waveform.shape
+            pixel_index = np.arange(n_pixels)
+
+            if selected_gain_channel is not None:
+                r1.waveform = waveform[selected_gain_channel, pixel_index]
+                r1.selected_gain_channel = selected_gain_channel
+            else:
+                r1.waveform = waveform
+                r1.selected_gain_channel = None
+
+            # store calibration data needed for dl1 calibration in ctapipe
+            # first drs4 time shift (zeros if no calib file was given)
+            time_shift = self.get_drs4_time_correction(
+                tel_id, self.first_cap[tel_id],
+                selected_gain_channel=selected_gain_channel,
+            )
+
+            # time shift from flat fielding
+            if self.mon_data is not None and self.add_calibration_timeshift:
+                time_corr = self.mon_data.tel[tel_id].calibration.time_correction
+                # time_shift is subtracted in ctapipe,
+                # but time_correction should be added
+                if selected_gain_channel is not None:
+                    time_shift -= time_corr[r1.selected_gain_channel, pixel_index].to_value(u.ns)
+                else:
+                    time_shift -= time_corr.to_value(u.ns)
+
+            event.calibration.tel[tel_id].dl1.time_shift = time_shift
+
+            # needed for charge scaling in ctpaipe dl1 calib
+            if selected_gain_channel is not None:
+                relative_factor = np.empty(n_pixels)
+                relative_factor[r1.selected_gain_channel == HIGH_GAIN] = self.calib_scale_high_gain.tel[tel_id]
+                relative_factor[r1.selected_gain_channel == LOW_GAIN] = self.calib_scale_low_gain.tel[tel_id]
+            else:
+                relative_factor = np.empty((n_gains, n_pixels))
+                relative_factor[HIGH_GAIN] = self.calib_scale_high_gain.tel[tel_id]
+                relative_factor[LOW_GAIN] = self.calib_scale_low_gain.tel[tel_id]
+
+            event.calibration.tel[tel_id].dl1.relative_factor = relative_factor
```

Comment:
Actually the scaling factors are not conceptually de relative factors, but indeed this solution will work ;-) 
I wonder here about the dl1.pedestal_offset, they are default equal None in the container, should we not better initialise them to zero in ctapipe (better) or here?

Reply:
If we don't want to use them, letting stay `None` will save memory and processing time, since not doing a subtractions is cheaper than subtracting 0

Reply:
But if are running the ctapipe-stage1 script we are not going through the usual calibration process?

Reply:
I don't understand what you mean by this.

Reply:
the ctapipe calibrator, e.g in stage1.py:


```
        for event in tqdm(
            self.event_source,
            desc=self.event_source.__class__.__name__,
            total=self.event_source.max_events,
            unit="ev",
            disable=not self.progress_bar,
        ):

            self.log.log(9, "Processessing event_id=%s", event.index.event_id)
            self.calibrate(event)
            if self.write_dl1.write_parameters:
                self.process_images(event)
            self.write_dl1(event)
```

Reply:
Yes exactly. So if we don't need to give an offset, its better to leave it None than to use an array of zeros.